csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +504,221 @@ class RedundantUseAnalysis : BackwardVisitor {
+        TORCH_INTERNAL_ASSERT(
+            outer_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->outer());
+        TORCH_INTERNAL_ASSERT(
+            inner_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->inner());
+        int dist = std::distance(inner_iter_im, outer_iter_im);
+        if (std::abs(dist) != 1) {
+          // if the merged domains are not adjacent, return empty so the
+          // predication is not applied.
+          return std::vector<IterDomain*>();
+        }
+        // insert after erase so the beg iter is not invalidated
+        auto beg = dist > 0 ? inner_iter_im : outer_iter_im;
+        auto end = dist > 0 ? outer_iter_im : inner_iter_im;
+        intermediate_domains.erase(beg, end + 1);
+        intermediate_domains.insert(beg, merge->out());
+
+        auto outer_iter =
+            std::find(root_domain_.begin(), root_domain_.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain_.begin(), root_domain_.end(), merge->inner());
+        // outer_iter and inner_iter should be neighbors in root_domain
+        if (inner_iter != root_domain_.end() &&
+            outer_iter != root_domain_.end() &&
+            std::distance(outer_iter, inner_iter) != 1) {
+          return std::vector<IterDomain*>();
+        }
+        if (outer_iter != root_domain_.end()) {
+          merged_root_domains.emplace_back(*outer_iter);
+          index_root_domain.emplace_back(
+              std::distance(root_domain_.begin(), outer_iter));
+        }
+        if (inner_iter != root_domain_.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(
+              std::distance(root_domain_.begin(), inner_iter));
+        }
+      } else {
+        // current analysis of predication is only valid if all the exprs
+        // between this lead domain and root domains are merge
+        return std::vector<IterDomain*>();
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [I3, I1, B2] while the correct order should be [I1, B2, I3]
+    int n_elements = merged_root_domains.size();
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain[a] < index_root_domain[b];
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (int i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted[i] = merged_root_domains[indices[i]];
+    }
+
+    return merged_root_domains_sorted;
+  }
+
+  // Get the index of the leaf domain if we skip the broadcasted root domains
+  Val* getIndexWithoutBroadcast(
+      const std::vector<IterDomain*>& merged_root_domains,
+      ParallelType pt) {
+    const int ndim = (int)merged_root_domains.size();
```

Comment:
YES. The reason is becuase every leaf domain must come from at least a root/rfactor domain. I added `TORCH_INTERNAL_ASSERT(n_elements, "The number of merged root domains should > 0");` right after the filling of `merged_root_domains` is done.

Reply:
Actually, I don't think so because `getRootDomainsMergedToLeaf` can return an empty vector at line 622, which is not caught by the assertion.

I strongly recommend `std::vector::at()` should be used almost all cases instead of `std::vector[]`. `at` does range check, so it's safer but slower than `[]`. However, the perf benefit of omitting the range change is almost all cases not justifiable considering the difficulty of debugging a memory corruption caused by `[]`.

Reply:
make sense. added protection, will move forward only if `(!merged_root_domains.empty()) `
change to std::vector::at() in `csrc/device_lower/analysis/thread_predicate.cpp`